### PR TITLE
DE-3343 IDS Migration Script

### PR DESCRIPTION
**Pull request by VladimirShteyn**
A helper script for converting IDS to PIDS. See the changes to the README.md for details.

Three reviewers were added. @robjhornby-ts or @nilsguillermin whoever has more time, please review. @mlipschultz and @spinwang are FYI. (Review if you have time.)

### CODE REVIEW

**Comment by nilsguillermin** [comment]
Ran into an error, posting traceback on request.

```(ts-ids-core-sTW2XApH-py3.7) nilsguillermin@Nilss-MacBook-Pro ts-ids-core % import-schema -i ../ts-lib-protocol-script/ids/common/plate-reader-msd-sector-s-600/v1.0.0/schema.json                                 [ts-ids-core/DE-3343-ids-migration-script]
Traceback (most recent call last):
  File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/__main__.py", line 553, in main
    use_non_positive_negative_number_constrained_types=config.use_non_positive_negative_number_constrained_types,
  File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/__init__.py", line 363, in generate
    results = parser.parse()
  File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/base.py", line 435, in parse
    self.parse_raw()
  File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 1189, in parse_raw
    self._parse_file(self.raw_obj, obj_name, path_parts)
  File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 1250, in _parse_file
    root_obj = JsonSchemaObject.parse_obj(raw)
  File "pydantic/main.py", line 511, in pydantic.main.BaseModel.parse_obj
  File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 199, in __init__
    super().__init__(**data)
  File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 9 validation errors for JsonSchemaObject
properties -> results -> items
  value is not a valid list (type=type_error.list)
properties -> results -> items -> properties -> sectors -> items
  value is not a valid list (type=type_error.list)
properties -> results -> items -> properties -> sectors -> items -> properties -> camera -> properties -> additionalProperties
  value is not a valid dict (type=type_error.dict)
properties -> results -> items -> properties -> sectors -> items -> properties -> additionalProperties
  value is not a valid dict (type=type_error.dict)
properties -> datacubes -> items
  value is not a valid list (type=type_error.list)
properties -> datacubes -> items -> properties -> dimensions -> items
  value is not a valid list (type=type_error.list)
properties -> datacubes -> items -> properties -> dimensions -> items -> properties -> additionalProperties
  value is not a valid dict (type=type_error.dict)
properties -> datacubes -> items -> properties -> additionalProperties
  value is not a valid dict (type=type_error.dict)
properties -> additionalProperties
  value is not a valid dict (type=type_error.dict)```

File: ts_ids_core/scripts/jsonschema_to_programmatic_ids.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    """See help strings for parameter documentation."""
+    input_abs_path = pathlib.Path(input_).absolute()
+    if not input_abs_path.exists():
+        raise FileNotFoundError(
+            f"The input file, {str(input_abs_path)}, does not exist."
+        )
+
+    if outfile is None:
+        to_dir = input_abs_path.parent
+        out = to_dir.joinpath(f"{input_abs_path.stem}.py")
```
Comment (nilsguillermin) [inline ts_ids_core/scripts/jsonschema_to_programmatic_ids.py:83]: bug here: `out` should be `outfile`.

**Review by nilsguillermin — APPROVED** [review]

**Comment by VladimirShteyn** [comment]
> Ran into an error, posting traceback on request.
> 
> ```
> Traceback (most recent call last):
>   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/__main__.py", line 553, in main
>     use_non_positive_negative_number_constrained_types=config.use_non_positive_negative_number_constrained_types,
>   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/__init__.py", line 363, in generate
>     results = parser.parse()
>   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/base.py", line 435, in parse
>     self.parse_raw()
>   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 1189, in parse_raw
>     self._parse_file(self.raw_obj, obj_name, path_parts)
>   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 1250, in _parse_file
>     root_obj = JsonSchemaObject.parse_obj(raw)
>   File "pydantic/main.py", line 511, in pydantic.main.BaseModel.parse_obj
>   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 199, in __init__
>     super().__init__(**data)
>   File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
> pydantic.error_wrappers.ValidationError: 9 validation errors for JsonSchemaObject
> properties -> results -> items
>   value is not a valid list (type=type_error.list)
> properties -> results -> items -> properties -> sectors -> items
>   value is not a valid list (type=type_error.list)
> properties -> results -> items -> properties -> sectors -> items -> properties -> camera -> properties -> additionalProperties
>   value is not a valid dict (type=type_error.dict)
> properties -> results -> items -> properties -> sectors -> items -> properties -> additionalProperties
>   value is not a valid dict (type=type_error.dict)
> properties -> datacubes -> items
>   value is not a valid list (type=type_error.list)
> properties -> datacubes -> items -> properties -> dimensions -> items
>   value is not a valid list (type=type_error.list)
> properties -> datacubes -> items -> properties -> dimensions -> items -> properties -> additionalProperties
>   value is not a valid dict (type=type_error.dict)
> properties -> datacubes -> items -> properties -> additionalProperties
>   value is not a valid dict (type=type_error.dict)
> properties -> additionalProperties
>   value is not a valid dict (type=type_error.dict)```
> ```

Adding repro steps from Slack messages:

```
cd ts-ids-core
poetry shell
import-schema -i ../ts-lib-protocol-script/ids/common/plate-reader-msd-sector-s-600/v1.0.0/schema.json
```

(Commit hash of of `ts-lib-protocol-script` is `bf7b02ea2cf45d6ce49befb3ff7065fc5a415aa2`).

File: ts_ids_core/scripts/jsonschema_to_programmatic_ids.py (hunk 1 above)
Comment (VladimirShteyn) [reply ts_ids_core/scripts/jsonschema_to_programmatic_ids.py]: Fixed.

**Comment by VladimirShteyn** [comment]
> > Ran into an error, posting traceback on request.
> > ```
> > Traceback (most recent call last):
> >   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/__main__.py", line 553, in main
> >     use_non_positive_negative_number_constrained_types=config.use_non_positive_negative_number_constrained_types,
> >   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/__init__.py", line 363, in generate
> >     results = parser.parse()
> >   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/base.py", line 435, in parse
> >     self.parse_raw()
> >   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 1189, in parse_raw
> >     self._parse_file(self.raw_obj, obj_name, path_parts)
> >   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 1250, in _parse_file
> >     root_obj = JsonSchemaObject.parse_obj(raw)
> >   File "pydantic/main.py", line 511, in pydantic.main.BaseModel.parse_obj
> >   File "/Users/nilsguillermin/Library/Caches/pypoetry/virtualenvs/ts-ids-core-sTW2XApH-py3.7/lib/python3.7/site-packages/datamodel_code_generator/parser/jsonschema.py", line 199, in __init__
> >     super().__init__(**data)
> >   File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
> > pydantic.error_wrappers.ValidationError: 9 validation errors for JsonSchemaObject
> > properties -> results -> items
> >   value is not a valid list (type=type_error.list)
> > properties -> results -> items -> properties -> sectors -> items
> >   value is not a valid list (type=type_error.list)
> > properties -> results -> items -> properties -> sectors -> items -> properties -> camera -> properties -> additionalProperties
> >   value is not a valid dict (type=type_error.dict)
> > properties -> results -> items -> properties -> sectors -> items -> properties -> additionalProperties
> >   value is not a valid dict (type=type_error.dict)
> > properties -> datacubes -> items
> >   value is not a valid list (type=type_error.list)
> > properties -> datacubes -> items -> properties -> dimensions -> items
> >   value is not a valid list (type=type_error.list)
> > properties -> datacubes -> items -> properties -> dimensions -> items -> properties -> additionalProperties
> >   value is not a valid dict (type=type_error.dict)
> > properties -> datacubes -> items -> properties -> additionalProperties
> >   value is not a valid dict (type=type_error.dict)
> > properties -> additionalProperties
> >   value is not a valid dict (type=type_error.dict)```
> > ```
> 
> Adding repro steps from Slack messages:
> 
> ```
> cd ts-ids-core
> poetry shell
> import-schema -i ../ts-lib-protocol-script/ids/common/plate-reader-msd-sector-s-600/v1.0.0/schema.json
> ```
> 
> (Commit hash of of `ts-lib-protocol-script` is `bf7b02ea2cf45d6ce49befb3ff7065fc5a415aa2`).

This error has been addressed in the [guide Confluence page](https://tetrascience.atlassian.net/wiki/spaces/DE/pages/3189213830/DE-3320+How+to+semi-automatically+convert+IDS+jsonschema+to+Programmatic+IDS).

**Pull request merged**